docs/proposals/20210310-opt-in-autoscaling-from-zero.md

```diff
@@ -0,0 +1,301 @@
+for the keys and `k8s.io/apimachinery/pkg/api/resource.Quantity` as the values (similar to how resource
+limits and requests are handled for pods).
+
+It is worth mentioning that the Infrastructure Machine Templates are not usually reconciled by themselves.
+Each infrastructure provider will be responsible for determining the best implementation for adding the
+status field based on the information available on their platform.
+
+**Example implementation in Docker provider**
+```
+// these constants will be carried in Cluster API, but are repeated here for clarity
+const (
+    AutoscalerResourceCPU ResourceName = "cpu"
+    AutoscalerResourceMemory ResourceName = "memory"
+    AutoscalerResourceGPU ResourceName = "gpu"
+)
+
+// AutoscalerMachineStatus defines information about individual machines used by the cluster autoscaler to perform scale from zero operations
+type AutoscalerMachineStatus struct {
+    Request ResourceList `json:"request,omitempty"`
+}
+
+// AutoscalerMachineTemplateStatus defines information that is used by the cluster autoscaler to perform scale from zero operations
+type AutoscalerMachineTemplateStatus struct {
+    Machine AutoscalerMachineStatus `json:"machine,omitempty"`
+}
+
+// DockerMachineTemplateStatus defines the observed state of a DockerMachineTemplate
+type DockerMachineTemplateStatus struct {
+    Autoscaler AutoscalerMachineTemplateStatus `json:"autoscaler,omitempty"`
+}
+
+// DockerMachineTemplate is the Schema for the dockermachinetemplates API.
+type DockerMachineTemplate struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    Spec DockerMachineTemplateSpec     `json:"spec,omitempty"`
+    Status DockerMachineTemplateStatus `json:"status,omitempty"`
+}
+```
+_Note: the `ResourceList` and `ResourceName` referenced are from k8s.io/kubernetes/pkg/apis/core/types`_
+
+When used as a manifest, it would look like this:
+
+```
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+kind: DockerMachineTemplate
+metadata:
+  name: workload-md-0
+  namespace: default
+spec:
+  template:
+    spec: {}
+status:
+  autoscaler:
+    machine:
+      request:
+        memory: 500mb
+        cpu: "1"
+        gpu: "1"
+```
+
+#### MachineSet and MachineDeployment Annotations
+
+In cases where a user needs to provide specific resource information for a
+MachineSet or MachineDeployment, or in cases where an infrastructure provider
+has not yet added the Infrastructure Machine Template status changes, they
+may use annotations to provide the information.
+
+If a user wishes to specify the resource requests through annotations, they
+may do by adding the following to any MachineSet or MachineDeployment (it is not required on both)
+that are participating in autoscaling:
+
+```
+kind: <MachineSet or MachineDeployment>
+metadata:
+  annotations:
+      autoscaler-machine-request.x-k8s.io/gpu: "1"
```

Comment:
Are those annotations already specified in the autoscaler? If not, should we use the Cluster API namespace?

Reply:
they are not, i was trying to suggest a name that would make it clear.  what if i change to:
```
cluster.x-k8s.io/autoscaler-machine-request-gpu
```

and if we like `capacity` better for the status block, i would change the annotation name accordingly.

Reply:
WDYT about: `autoscaler.cluster.x-k8s.io/machine-request-gpu`?
But I'm fine with whatever you can get consensus on :)

I just think we probably shouldn't use `x-k8s.io` without further namespacing.

Reply:
makes sense to me, i will wait until we decide the other naming question and then fix this as well. thanks!